File: Backend/src/main/kotlin/com/backend/is/IS_Backend/service/impl/AuthServiceImpl.java
package com.backend.is.IS_Backend.service.impl;

import mk.ukim.finki.it.reservengo.dto.jwtDTO.JWTAuthenticationRequestDTO;
import mk.ukim.finki.it.reservengo.dto.jwtDTO.JWTAuthenticationResponseDTO;
import mk.ukim.finki.it.reservengo.dto.jwtDTO.JWTLoginDTO;
import mk.ukim.finki.it.reservengo.model.domain.Customer;
import mk.ukim.finki.it.reservengo.model.domain.LocalManager;
import mk.ukim.finki.it.reservengo.model.domain.LocalWorker;
import mk.ukim.finki.it.reservengo.model.domain.User;
import mk.ukim.finki.it.reservengo.model.exceptions.EmailNotFoundException;
import mk.ukim.finki.it.reservengo.repository.UserRepository;
import mk.ukim.finki.it.reservengo.service.intf.*;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
public class AuthServiceImpl implements AuthService {
    private final CustomerService customerService;
    private final LocalWorkerService localWorkerService;
    private final LocalManagerService localManagerService;
    private final JWTService jwtService;
    private final PasswordEncoder passwordEncoder;
    private final AuthenticationManager authenticationManager;
    private final UserRepository userRepository;
    private final UserService userService;

    public AuthServiceImpl(CustomerService customerService, LocalWorkerService localWorkerService, LocalManagerService localManagerService, JWTService jwtService, PasswordEncoder passwordEncoder, AuthenticationManager authenticationManager, UserRepository userRepository, UserService userService) {
        this.customerService = customerService;
        this.localWorkerService = localWorkerService;
        this.localManagerService = localManagerService;
        this.jwtService = jwtService;
        this.passwordEncoder = passwordEncoder;
        this.authenticationManager = authenticationManager;
        this.userRepository = userRepository;
        this.userService = userService;
    }

    @Override
    public JWTAuthenticationResponseDTO registerCustomer(JWTAuthenticationRequestDTO jwtAuthenticationRequestDTO) {
        Customer customer = jwtAuthenticationRequestDTO.toCustomer(passwordEncoder.encode(jwtAuthenticationRequestDTO.password()));

        customerService.save(customer);
        String jwt = jwtService.generateToken(customer);

        return JWTAuthenticationResponseDTO.fromUser(customer, jwt);
    }

    @Override
    public JWTAuthenticationResponseDTO registerLocalWorker(JWTAuthenticationRequestDTO jwtAuthenticationRequestDTO, String token) {
        validateTokenEmail(token, jwtAuthenticationRequestDTO.email());

        LocalWorker localWorker = jwtAuthenticationRequestDTO.toLocalWorker(passwordEncoder.encode(jwtAuthenticationRequestDTO.password()));

        localWorkerService.save(localWorker);
        String jwt = jwtService.generateToken(localWorker);

        return JWTAuthenticationResponseDTO.fromUser(localWorker, jwt);
    }

    @Override
    public JWTAuthenticationResponseDTO registerLocalManager(JWTAuthenticationRequestDTO jwtAuthenticationRequestDTO, String token) {
        validateTokenEmail(token, jwtAuthenticationRequestDTO.email());

        LocalManager localManager = jwtAuthenticationRequestDTO.toLocalManager(passwordEncoder.encode(jwtAuthenticationRequestDTO.password()));

        localManagerService.save(localManager);
        String jwt = jwtService.generateToken(localManager);

        return JWTAuthenticationResponseDTO.fromUser(localManager, jwt);
    }

    @Override
    public JWTAuthenticationResponseDTO login(JWTLoginDTO jwtLoginDTO) {
        authenticationManager.authenticate(new UsernamePasswordAuthenticationToken(jwtLoginDTO.email(), jwtLoginDTO.password()));

        User user = userRepository.findByEmail(jwtLoginDTO.email()).orElseThrow(() -> new EmailNotFoundException(jwtLoginDTO.email()));
        String jwt = jwtService.generateToken(user);

        userService.updateUserActivity(user.getId());

        return JWTAuthenticationResponseDTO.fromUser(user, jwt);
    }

    @Override
    public void reactivateProfile(JWTLoginDTO jwtLoginDTO) {
        User user = userService.findUserByEmail(jwtLoginDTO.email());

        if (user == null) {
            throw new EmailNotFoundException(jwtLoginDTO.email());
        }

        if (!passwordEncoder.matches(jwtLoginDTO.password(), user.getPassword())) {
            throw new BadCredentialsException("Bad credentials.") ;
        }

        if (!user.isEnabled()) {
            userService.enableProfile(user.getId());
        }
    }

    private void validateTokenEmail(String token, String email) {
        String tokenEmail = jwtService.extractClaim(token, claims -> claims.get("email", String.class));
        if (!tokenEmail.equals(email)) {
            throw new IllegalArgumentException("Email must match the token's email");
        }
    }
}

File: Backend/src/main/kotlin/com/backend/is/IS_Backend/service/impl/AuthServiceImpl.kt
package com.backend.`is`.IS_Backend.service.impl

import com.backend.`is`.IS_Backend.dto.auth.AuthRequestDTO
import com.backend.`is`.IS_Backend.dto.auth.CountResponseDTO
import com.backend.`is`.IS_Backend.dto.auth.LoginDTO
import com.backend.`is`.IS_Backend.enumerations.Roles
import com.backend.`is`.IS_Backend.exception.EmailAlreadyInUseException
import com.backend.`is`.IS_Backend.exception.InvalidEmailFormatException
import com.backend.`is`.IS_Backend.exception.WeakPasswordException
import com.backend.`is`.IS_Backend.model.domain.Admin
import com.backend.`is`.IS_Backend.model.domain.Librarian
import com.backend.`is`.IS_Backend.model.domain.Reader
import com.backend.`is`.IS_Backend.model.domain.User
import com.backend.`is`.IS_Backend.repository.UserRepository
import com.backend.`is`.IS_Backend.service.intf.AuthServiceKt
import com.backend.`is`.IS_Backend.service.intf.JWTService
import com.backend.`is`.IS_Backend.service.intf.TwoFactorAuthService
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.concurrent.ConcurrentHashMap

@Service
class AuthServiceImpl(
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder,
    private val authenticationManager: AuthenticationManager,
    private val jwtService: JWTService,
    private val twoFactorAuthService: TwoFactorAuthService
) : AuthServiceKt {

    val EMAIL_REGEX = Regex("^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}$")
    private val STRONG_PASSWORD_REGEX = Regex("^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)(?=.*[^A-Za-z\\d]).{8,}$")

    private val pendingRegistrations = ConcurrentHashMap<String, User>()

    private fun validateCredentials(email: String, password: String) {
        if (!EMAIL_REGEX.matches(email)) {
            throw InvalidEmailFormatException("Email must be a valid email address")
        }
        if (!STRONG_PASSWORD_REGEX.matches(password)) {
            throw WeakPasswordException("Password must be at least 8 characters long and include uppercase, lowercase, number, and special character")
        }
        if (userRepository.findByEmail(email).isPresent || pendingRegistrations.containsKey(email)) {
            throw EmailAlreadyInUseException()
        }
    }

    @Transactional
    override fun registerAdmin(request: AuthRequestDTO) {
        validateCredentials(request.email, request.password)
        val user = Admin(request.email, passwordEncoder.encode(request.password))
        pendingRegistrations[request.email] = user
        twoFactorAuthService.sendCode(request.email)
    }

    @Transactional
    override fun registerLibrarian(request: AuthRequestDTO) {
        validateCredentials(request.email, request.password)
        val user = Librarian(request.email, passwordEncoder.encode(request.password))
        pendingRegistrations[request.email] = user
        twoFactorAuthService.sendCode(request.email)
    }

    @Transactional
    override fun registerReader(request: AuthRequestDTO) {
        validateCredentials(request.email, request.password)
        val user = Reader(request.email, passwordEncoder.encode(request.password))
        pendingRegistrations[request.email] = user
        twoFactorAuthService.sendCode(request.email)
    }

    @Transactional
    override fun verifyRegister(email: String, code: Int) {
        val pending = pendingRegistrations[email] ?: throw IllegalArgumentException("No pending registration for email")
        val ok = twoFactorAuthService.verifyCodeOnly(email, code)
        if (!ok) throw IllegalArgumentException("Invalid code")
        userRepository.save(pending)
        pendingRegistrations.remove(email)
    }

    override fun login(request: LoginDTO) : String {
        authenticationManager.authenticate(
            UsernamePasswordAuthenticationToken(request.email, request.password)
        )
        val user = userRepository.findByEmail(request.email).orElseThrow {
            IllegalArgumentException("Invalid credentials")
        }

        twoFactorAuthService.sendCode(request.email)
        return jwtService.generateToken(user)
    }
    override fun makeCount(user: User) : CountResponseDTO {
        user.setCount(user.getCount() + 1)
        userRepository.save(user)
        return CountResponseDTO(user.getCount());
    }

    override fun getCount(user: User) : CountResponseDTO {
        return CountResponseDTO(user.getCount())
    }
}